[21.05] fc-ceph-load-vm-images: fetch image list from directory

osnyx opened the pull request:
@flyingcircusio/release-managers

## Release process

depends on directory API change https://gitlab.flyingcircus.io/flyingcircus/fc.directory/-/merge_requests/455
Otherwise, image loading stops working.

Also includes PL-132048

Impact: none

internal only; breaks ceph image loading without corresponding directory change; directory is released and has been tested in dev

Changelog:

- Extend our VM base loading process to automatically pick up all known environments instead of using a hard coded list. (PL-132516)

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - [x] must not introduce known regressions
  - [x] extend test coverage
- [x] Security requirements tested? (EVIDENCE)
  - [x] NixOS and unit tests still pass
  - [x] newly introduced functions covered with unit tests
  - [x] added test for continuing release data parsing at errors
  - [x] manual integration test in dev environment with broken API data
  - [x] manual successful end-to-end test in dev with new directory API:
    - deliberately not attempted to fully cover existing legacy code with unit tests in one go

```
root@cartman06 /etc/nixos # fc-ceph maintenance load-vm-images
Updating environment fc-23.11-production
$ rbd --format json ls rbd.hdd
$ rbd lock add rbd.hdd/fc-23.11-production cartman06.295757
$ rbd --format json lock ls rbd.hdd/fc-23.11-production
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
        Have builds:
                build-450214
                build-461864
                build-472672
        Downloading build: 2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
        Storing in volume rbd.hdd/fc-23.11-production
$ rbd --format json info rbd.hdd/fc-23.11-production
$ rbd map rbd.hdd/fc-23.11-production
$ rbd unmap /dev/rbd0
        Creating snapshot 2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
$ rbd snap create rbd.hdd/fc-23.11-production@2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
$ rbd snap protect rbd.hdd/fc-23.11-production@2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
$ rbd --format json children rbd.hdd/fc-23.11-production@build-450214
$ rbd --format json children rbd.hdd/fc-23.11-production@build-461864
$ rbd --format json children rbd.hdd/fc-23.11-production@build-472672
$ rbd --format json children rbd.hdd/fc-23.11-production@2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
        Purging snapshot rbd.hdd/fc-23.11-production@build-450214
$ rbd snap unprotect rbd.hdd/fc-23.11-production@build-450214
$ rbd snap rm rbd.hdd/fc-23.11-production@build-450214
$ rbd lock remove rbd.hdd/fc-23.11-production cartman06.295757 client.3490297474

root@cartman06 /etc/nixos # fc-ceph maintenance load-vm-images
Updating environment fc-23.11-production
$ rbd --format json ls rbd.hdd
$ rbd lock add rbd.hdd/fc-23.11-production cartman06.296498
$ rbd --format json lock ls rbd.hdd/fc-23.11-production
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
$ rbd --format json children rbd.hdd/fc-23.11-production@build-461864
$ rbd --format json children rbd.hdd/fc-23.11-production@build-472672
$ rbd --format json children rbd.hdd/fc-23.11-production@2024_025-5d572c28df4862cf8493a0bd2aabb3c9dc11893c340955a8131a67eef321e17c
$ rbd --format json snap ls rbd.hdd/fc-23.11-production
$ rbd lock remove rbd.hdd/fc-23.11-production cartman06.296498 client.3490297681

root@cartman06 /etc/nixos # fc-ceph maintenance load-vm-images --name-glob "lol"

root@cartman06 /etc/nixos # fc-ceph maintenance load-vm-images --name-glob "lol*"
```

osnyx committed:
vendor fc.util.directory from agent to fc-ceph code

osnyx committed:
fc-ceph maintenance load-vm-images: pull from directory

The fc-ceph image loading job now pulls the list of images to process
from the directory API (needs directory changes). Which environments to
process is determined by a name glob that can be customised via a cli
argument.

`fc-ceph maintenance load-vm-images --name-glob 'some-env*'` allows
specifying which environment image(s) to pull as a manual one-off
action.

Querying Hydra directly is removed from the fc-ceph code, as this is now
done by the directory anyways.

PL-132516

osnyx committed:
tests for sri<->sha256 hash conversion

osnyx committed:
fc-ceph maintenance load-vm-images: tests for data processing

osnyx committed:
test that image data parsing continues at errors

PL-132048

ctheune commented on pkgs/fc/ceph/src/fc/ceph/maintenance/images_nautilus.py:
```diff
@@ -13,134 +15,93 @@
+    "pre-fc-*",
+    "ts-test",
+]
+
```
Lets remove this globbing. If we need filtering, I'd rather delegate that to a feature in the directory at some point.

ctheune reviewed (changes requested):
Generally happy. Remove the globbing and just take all images.

osnyx committed:
fc-ceph-load-vm-images: remove unnecessery sleeps

osnyx committed:
fc-ceph-load-vm-images: rework image cleanup

In the ideal case, only the single current snapshot of an image is kept.
In case of a race condition, we try again at the next run.

Always keep at least one snapshot as a safety measure against invalid
image list data.

suggestion how to make the code a bit easier to read

osnyx commented:
Re-worked the image purging. Only the single most-current snapshot is left in the ideal case; race-conditions are accounted for with retrying at the next run.

osnyx commented on pkgs/fc/ceph/src/fc/ceph/tests/test_images.py:
```diff
@@ -0,0 +1,142 @@
+
+import pytest
+
+# TODO validate data format by calling dev deployment’s directory API
```
this is done

ctheune commented on pkgs/fc/ceph/src/fc/ceph/maintenance/images_nautilus.py:
```diff
@@ -353,37 +326,16 @@ def flatten(self):
                 run.rbd("snap", "rm", snap_spec)
             except Exception:
-                logger.exception("Error trying to purge snapshot:")
+                logger.exception(
```
wenn unprotect schief geht und rm nicht, dann ist das jetzt nicht mehr konvergent.

ctheune commented on pkgs/fc/ceph/src/fc/ceph/maintenance/images_nautilus.py:
```diff
@@ -273,72 +232,86 @@ def store_in_ceph(self, img):
+                continue
+
+            # safety belt: don't purge the only remaining snapshot
+            if num_snaps == 1:
```
num_snaps wird vor der schleife einmal ausgerechnet, das funktioniert so nicht

osnyx replied:
Doch, dass soll so: Wenn vor Beginn bereits feststeht, dass nur 1 snapshot insgesamt für das image vorhanden ist, wird hier abgebrochen. Der 1 Snapshot soll erhalten bleiben.

ctheune reviewed (changes requested):
ein wichtiger und ein kleinerer bug haben sich da noch eingeschlichen

osnyx commented on pkgs/fc/ceph/src/fc/ceph/maintenance/images_nautilus.py:
```diff
@@ -273,75 +232,91 @@ def store_in_ceph(self, img):
+        snap_specs = [self.volume + "@" + snap["name"] for snap in snaps]
+
+        # hard invariant: always keep at least 1 snapshot per image
+        if len(snap_specs) == 1:
```
Adopted the proposed changes with a slight change: I accept an empty list of snaps here, so that the assertion about a missing current snapshot is still triggered later in that case. This is a case we should still be notified about.
